### PR TITLE
User/avinp/generate turn api

### PR DESCRIPTION
**Pull request by apattath**
This PR updates the Azure.Communication.Identity library with an additional API GetTurnCredentials.
This endpoint in the service is gated behind an AFEC and enabled for a preview version.
This version of the SDK is intended to be generated only in private preview at this time and released through the dev feed (not the public nuget feed). Hence, the suggestion from Wes Haggard was to submit this to a feature branch and release SDK from the feature branch into the dev feed.

Link to PR with updated swagger: https://github.com/Azure/azure-rest-api-specs/pull/12881

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs:62]: Does this contain multiple urls?

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:311]: Why not use the `RestClient` that is instantiated in the constructor? It should not be instantiating it every time.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:331]: summary should be updated

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:77]: Your baseline code is out of date, you should probably rebase onto latest master

File: sdk/communication/Azure.Communication.Identity/src/autorest.md
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/src/autorest.md:11]: Did not you merge the swagger files? You should not need to import them both here

File: sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs:73]: Do we not support managed identity?

File: sdk/communication/Azure.Communication.Identity/src/autorest.md
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/autorest.md:11]: I did not merge the swagger files. As I mentioned in our offline chat, I created a preview version of the CommunicationIdentity file with the same preview version as that of the TURN Swagger. Since the TURN Swagger was in a separate folder/location, I didn't move it.

File: sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs:73]: TURN endpoint only supports HMAC auth, doesn't support BearerToken Authentication. The Managed Identity test in the SDK was sending the token as bearer token and I was getting 'unauthorized'.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:77]: Yes, I think there were some changes overnight, which I didn't get a chance to merge. I will send a new iteration with merged changes.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:311]: I see this as temporary until the Rest endpoint version matches that of the identity service. Currently both are at different versions, hence need separate clients.

File: sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs:62]: It doesn't. I had discussed this with Andrew (@ajpeacock0 ) about this few days ago. It looks like according to the [WebRTC standard, `url` is deprecated.](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer#obsolete_properties)

Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs:62]: Oh nice, it is good that we are following the standards. One observation thought, [urls](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer/urls) could be an array as well. 
```
urls = iceServerUrl | [ url1, ..., urlN ]
```
which unfortunately not easy to do in swagger. But If we switch to `array` at least we will be future proof

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:77]: There are some old changes as well around how the client gets constructed

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:315]: is not turn on a preview version?

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:311]: Maybe we should have two members `RestClient` and `TurnRestClient` both created in the constructor. Let's make sure the new preview api also support all the `2021-03-07` routes as this looks hacky.

File: sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs:73]: What is the plan for supporting managed identity, because `CommunicationIdentityClient` can be instantiated using managed identity

File: sdk/communication/Azure.Communication.Common/tests/CommunicationRecordedTestSanitizer.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/CommunicationRecordedTestSanitizer.cs:15]: With the [change from earlier today](https://github.com/Azure/azure-sdk-for-net/pull/18715), this should be move to the sanitizer from the identity package.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:315]: Not yet. The change to move the TURN service endpoint to a preview version was checked-in yesterday (it's in Dev/Int, but not in PROD yet). So, I need to coordinate this based on when the service change will get deployed to PROD.

File: sdk/communication/Azure.Communication.Common/tests/CommunicationRecordedTestSanitizer.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Common/tests/CommunicationRecordedTestSanitizer.cs]: Updated to follow the new sanitizer rules.

File: sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs:73]: Looking into this one. Apparently it's supported, but I get 401 unauthorized from the endpoint. Following up if there's something I'm missing with respect to scopes? The TURN endpoint itself uses AuthenticationScheme.Server, which means it should accept both HMAC and AAD auth.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs]: Merged with latest changes.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs]: Updated.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs]: Updated.

File: sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs:62]: Will follow up offline with you and Andrew.

File: sdk/communication/Azure.Communication.Identity/tests/Infrastructure/CommunicationIdentityClientRecordedTestSanitizer.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/tests/Infrastructure/CommunicationIdentityClientRecordedTestSanitizer.cs:11]: does url contain PII information? if not, we do not need to sanitize it. Same for username, is it just an id?

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:85]: Next time that we update the swagger, let's make sure the preview api contains both TURN and Identity routes to avoid having two clients

File: sdk/communication/Azure.Communication.Identity/tests/Infrastructure/CommunicationIdentityClientRecordedTestSanitizer.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/tests/Infrastructure/CommunicationIdentityClientRecordedTestSanitizer.cs:11]: No, it's not PII. It's just a static url (common for every client). username is just a random id it looks like, yes.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:85]: Discussing this offline with Reza. There were some concerns in just copying all of identity routes to a preview version without any effective changes.

File: sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/api/Azure.Communication.Identity.netstandard2.0.cs:62]: Talked to Reza and Andrew offline, will address this in a subsequent PR as this needs to be coordinated with both swagger and service side updates.

File: sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/tests/CommunicationIdentityClient/CommunicationIdentityClientLiveTests.cs:73]: Talked to Reza and Andrew offline, will address this in a subsequent PR as this needs to be coordinated with service side updates.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs]: This will also be updated after the service with preview version is deployed

File: sdk/communication/Azure.Communication.Identity/tests/Infrastructure/CommunicationIdentityClientRecordedTestSanitizer.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/tests/Infrastructure/CommunicationIdentityClientRecordedTestSanitizer.cs]: Updated.

File: sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs
Comment (apattath) [reply sdk/communication/Azure.Communication.Identity/src/CommunicationIdentityClient.cs:85]: This will be addressed (based on outcome of discussion) in a subsequent PR.

**Review by jsquire — APPROVED** [review]
I left a few thoughts; nothing that I'd consider blocking.   As I don't have familiarity with the service, I'll defer to @RezaJooyandeh for final approval.   I'm am happy to serve as the merge-button-preeser, if one is needed.

**Review by RezaJooyandeh — APPROVED** [review]
Ready to merge once you applied @jsquire 's comments.

**Pull request merged**
